aten/src/ATen/native/quantized/cpu/qfc.cpp

```diff
@@ -0,0 +1,152 @@
+#include <ATen/fbgemm_utils.h>
+
+#include <algorithm>
+#include <tuple>
+
+namespace at {
+namespace native {
+namespace {
+
+class QFCInt8 final : public c10::OperatorKernel {
+ public:
+#ifdef USE_FBGEMM
+  std::tuple<at::Tensor, double, int64_t> operator()(
+      const at::Tensor& input,
+      double input_scale,
+      int64_t input_zero_point,
+      const at::Tensor& packed_weight,
+      double weight_scale,
+      int64_t weight_zero_point,
+      const at::Tensor& bias,
+      double output_scale,
+      int64_t output_zero_point) {
+    // uint8 * int8 -> uint8 (no quantization/dequantization)
+
+    // We make a strong guarantee that models using these operators will have
+    // the same numerics across different machines. Therefore, we do not provide
+    // a fallback path and rather fail loudly if we cannot run FBGEMM.
+    AT_ASSERTM(
+        fbgemm::fbgemmSupportedCPU(), "Your CPU does not support FBGEMM.");
+
+    auto input_contig = input.contiguous();
+    const auto* input_ptr = input_contig.data<uint8_t>();
+
+    AT_ASSERT(input.dim() >= 2);
+    int64_t M = 1;
+    for (size_t i = 0; i < input.dim() - 1; ++i) {
+      M *= input.size(i);
+    }
+
+    // Pull out the PackBMatrix and col_offsets instance from the owning tensor
+    auto& pack_ptr = cpp_custom_type_hack::cast<PackedFCWeight>(packed_weight);
+    auto packB = pack_ptr.w.get();
+    // packB->printPackedMatrix("packedB inside qfc: ");
+    auto& col_offsets = pack_ptr.col_offsets;
+
+    int64_t N = static_cast<int64_t>(packB->numCols());
+    int64_t K = input.size(input.dim() - 1);
+    AT_ASSERT(K == static_cast<int64_t>(packB->numRows()));
+    AT_ASSERT(bias.size(0) == N);
+    AT_ASSERT(bias.dim() == 1);
+
+    float input_scale_float = static_cast<float>(input_scale);
+    int32_t input_zero_point_int32 = static_cast<int32_t>(input_zero_point);
+
+    float weight_scale_float = static_cast<float>(weight_scale);
+    int32_t weight_zero_point_int32 = static_cast<int32_t>(weight_zero_point);
+
+    float output_scale_float = static_cast<float>(output_scale) /
+        input_scale_float / weight_scale_float;
+    int32_t output_zero_point_int32 = static_cast<int32_t>(output_zero_point);
+
+    // This operation does the following:
+    // 1) Quantizes the input matrix given the statistics we've calculated above
+    // 2) Creates a "row buffer" vector with offset values that must be added
+    //    to the integer matrix multiplication operation to ensure correctness
+    // 3) Packs the resulting quantized matrix into vector-register and cache
+    //    friendly tiles.
+    //
+    //  Note this is not executed eagerly, but rather within the fbgemmPacked
+    //  call below.
+    fbgemm::PackAWithRowOffset<uint8_t> packA(
+        /*trans=*/fbgemm::matrix_op_t::NoTranspose,
+        /*nRow=*/M,
+        /*nCol=*/K,
+        /*smat=*/input_ptr,
+        /*ld=*/K,
+        /*pmat=*/nullptr); // packA manages ownership of `pmat`
```

Comment:
Maybe add a TODO to consider a way to pre-allocate (and reuse) pmat buffer?

Reply:
Will add this TODO.